src/tools/log-parser-rs/src/log_message.rs

```diff
@@ -0,0 +1,133 @@
+use std::fmt::Display;
+
+use chrono::{DateTime, Utc};
+use serde::{de::Visitor, Deserialize, Serialize};
+use serde_with::{serde_as, DisplayFromStr};
+use thiserror::Error;
+
+#[serde_as]
+#[derive(Serialize, Deserialize, Debug)]
+pub(crate) struct LogMessage {
+    level: LogLevel,
+
+    msg: String,
+
+    name: String,
+
+    #[serde_as(as = "DisplayFromStr")]
+    pid: usize,
+
+    source: String,
+
+    subsystem: String,
+
+    #[serde_as(as = "DisplayFromStr")]
+    #[serde(alias = "time", alias = "ts")]
+    timestamp: DateTime<Utc>,
+}
+
+//totally abusing serde to easily display this.
+impl Display for LogMessage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", serde_json::to_string_pretty(&self).unwrap())
+    }
+}
+
+// A newtype for slog::Level, as it does not implement Serialize and Deserialize.
+#[derive(Debug)]
+pub(crate) struct LogLevel(slog::Level);
+
+#[derive(Debug, Error)]
+pub enum LevelError {
+    #[error("invalid slog level: {0}")]
+    InvalidLevel(String),
+}
+
+impl TryFrom<&str> for LogLevel {
+    type Error = LevelError;
+
+    fn try_from(s: &str) -> Result<Self, LevelError> {
+        let level = match s.to_lowercase().as_str() {
+            //for some reason, some of the error messages contain partial strings.
```

Comment:
Could you give an example of this in a comment here?

Reply:
turns out they were not partial strings, but the 'short' version of the levels.